Lib/ufo2ft/featureWriters/baseFeatureWriter.py

```diff
@@ -119,6 +153,110 @@ def _write(self):
+        marker comment.
+
+        If the insert marker is at the top of a feature block, the feature is
+        inserted before that block, and after if the insert marker is at the
+        bottom.
+        If the insert marker is in the middle of a feature block, that block
+        is split in to two blocks separated by the feature.
+        """
+
+        statements = feaFile.statements
+
+        # Collect insert markers in blocks
+        insertComments = self.context.insertComments
+
+        wroteOthers = False
+        for _, feature in features:
+            if insertComments and feature.name in insertComments:
+                block, comment = insertComments[feature.name]
+                markerIndex = block.statements.index(comment)
+
+                onlyCommentsBefore = all(
+                    isinstance(s, ast.Comment) for s in block.statements[:markerIndex]
+                )
+                onlyCommentsAfter = all(
+                    isinstance(s, ast.Comment) for s in block.statements[markerIndex:]
+                )
+
+                # Remove insert marker(s) from feature block.
+                del block.statements[markerIndex]
+
+                # insertFeatureMarker is in a block with only comments.
+                # Replace that block with new feature block.
+                if onlyCommentsBefore and onlyCommentsAfter:
+                    index = statements.index(block)
+                    statements.remove(block)
+
+                # insertFeatureMarker is at the top of a feature block
+                # or only preceded by other comments.
+                elif onlyCommentsBefore:
+                    index = statements.index(block)
+
+                # insertFeatureMarker is at the bottom of a feature block
+                # or only followed by other comments
+                elif onlyCommentsAfter:
+                    index = statements.index(block) + 1
+
+                # insertFeatureMarker is in the middle of a feature block
+                # preceded and followed by statements that are not comments
+                #
+                # Glyphs3 can insert a feature block when rules are before
+                # and after the insert marker.
+                # See
+                # https://github.com/googlefonts/ufo2ft/issues/351#issuecomment-765294436
+                # This is currently not supported.
+                else:
+                    raise InvalidFeaturesData(
+                        "Insert marker has rules before and after, feature %s "
+                        "cannot be inserted." % block.name
+                    )
+
+            else:
+                index = len(statements)
+
+            if not wroteOthers:
+                others = []
+                # Insert classDefs
+                # Note: The AFDKO spec says glyph classes should be defined before any
+                # are used, but ufo2ft featureWriters has never done that.
```

Comment:
Where's that written? i understand them being defined before they are used, but defining all of them before any one is used is unnecessary limitation. Maybe we should raise that with the afdko spec

Reply:
I'm probably reading the spec to narrowly.